src/Raven.Server/Indexing/TempFileCache.cs

```diff
@@ -0,0 +1,156 @@
+using System;
+using System.Collections.Concurrent;
+using System.IO;
+using Raven.Server.ServerWide;
+using Raven.Server.Utils;
+using Sparrow.Global;
+using Sparrow.LowMemory;
+using Sparrow.Utils;
+using Voron;
+
+namespace Raven.Server.Indexing
+{
+    /// <summary>
+    /// This assume single threaded and is meant to be used for indexing only
+    /// We need to handle multi threaded from the low memory notifications, though 
+    /// </summary>
+    public sealed class TempFileCache : IDisposable, ILowMemoryHandler
+    {
+        private readonly StorageEnvironmentOptions _options;
+        private readonly ConcurrentQueue<FileStream> _files = new ConcurrentQueue<FileStream>();
+        private readonly ConcurrentQueue<MemoryStream> _ms = new ConcurrentQueue<MemoryStream>();
+
+        private const long MaxFileSizeToReduceInBytes = 16 * Constants.Size.Megabyte;
+        private const int MaxFilesToKeepInCache = 32;
+
+        public TempFileCache(StorageEnvironmentOptions options)
+        {
+            _options = options;
+            string path = _options.TempPath.FullPath;
+            if (Directory.Exists(path) == false)
+                Directory.CreateDirectory(path);
+            foreach (string file in Directory.GetDirectories(path,  "lucene-*" + StorageEnvironmentOptions.DirectoryStorageEnvironmentOptions.TempFileExtension))
```

Comment:
There shouldn't be any such temp files at this point because we delete them on storage startup. See https://github.com/ravendb/ravendb/pull/11497/files (https://issues.hibernatingrhinos.com/issue/RavenDB-15911)

Reply:
Yes, we don't expect that. However, I'm including this here in case the server crashed.
If that happened, we don't want the old files to still remain there (nothing will remove them).
For that purpose, we scan and remove / reuse them as needed.